src/planscape/collaboration/utils.py

```diff
@@ -0,0 +1,118 @@
+from collaboration.models import Collaborator, Permissions, Role
+from planning.models import PlanningArea, Scenario
+from django.contrib.auth import get_user_model
+from django.contrib.contenttypes.models import ContentType
+
+
+User = get_user_model()
+
+
+def is_creator(user: User, planning_area: PlanningArea):
+    return planning_area.user.pk == user.pk
+
+
+def is_owner(collaborator: Collaborator):
+    return collaborator.role == Role.OWNER
+
+
+def is_collaborator(collaborator: Collaborator):
+    return collaborator.role == Role.COLLABORATOR
+
+
+def is_viewer(collaborator: Collaborator):
+    return collaborator.role == Role.VIEWER
+
+
+def get_collaborator(user_id, planning_area_id):
+    planning_area_type = ContentType.objects.get(
+        app_label="planning", model="planningarea"
+    )
+
+    return Collaborator.objects.get(
+        collaborator_id=user_id,
+        content_type_id=planning_area_type,
+        object_pk=planning_area_id,
+    )
+
+
+def can_view_planning_area(user: User, planning_area: PlanningArea):
+    if is_creator(user, planning_area):
+        return True
+
+    try:
+        entry = get_collaborator(user.id, planning_area.id)
+        return is_viewer(entry) or is_collaborator(entry) or is_owner(entry)
+    except Collaborator.DoesNotExist:
+        return False
+
+
+def can_view_scenario(user: User, planning_area: PlanningArea):
+    if is_creator(user, planning_area):
+        return True
+
+    try:
+        entry = get_collaborator(user.id, planning_area.id)
+        return is_viewer(entry) or is_collaborator(entry) or is_owner(entry)
+    except Collaborator.DoesNotExist:
+        return False
+
+
+def can_add_scenario(user: User, planning_area: PlanningArea):
+    if is_creator(user, planning_area):
+        return True
+    try:
+        entry = get_collaborator(user.id, planning_area.id)
+        return is_collaborator(entry) or is_owner(entry)
```

Comment:
The more I think about it, testing for each role in these functions will get hard to maintain if we ever add *another* role type or remove one, or if we want to change which roles inherit what permissions. It won't be fun on the database either, but it would mean less risk or code changes.

That said, I couldn't cleanly get all of this to work with a single query, but we could  just get the permissions for this using something like:

`permissions = Permissions.objects.filter(role=entry)` then test for the specific permission, 
or alternatively we could just do something like: `Permissions.objects.filter(role=entry, permission='add_scenario')` to test for the specific value.

Reply:
Yeah! that was my original attempt, something like Permissions.objects.filter(role=entry, permission='add_scenario') 
but also joined with `Collaborators`. However we can probably do this with two queries.